### PR TITLE
Input validation tests

### PR DESCRIPTION
**Pull request by ahoffer**
Validate input from HTTP requests

### CODE REVIEW

File: src/main/java/com/connexta/store/controllers/StoreController.java [hunk 1]
```diff
@@ -6,6 +6,7 @@
  */
 package com.connexta.store.controllers;
 
+import static org.springframework.http.HttpStatus.*;
```
Comment (ahoffer) [inline src/main/java/com/connexta/store/controllers/StoreController.java]: There is something about IntelliJ converting imports to asterisks. Why doesn't Spotless doesn't fix these * imports?  If is bugs people I can go back and use explicit imports.

Comment (emmberk) [reply src/main/java/com/connexta/store/controllers/StoreController.java]: I think we should use explicit imports.

Comment (emmberk) [reply src/main/java/com/connexta/store/controllers/StoreController.java]: Here are my settings in Intellij
image

Comment (figliold) [reply src/main/java/com/connexta/store/controllers/StoreController.java]: I believe the company coding standards state using explicit imports.

File: src/main/java/com/connexta/store/controllers/StoreController.java [hunk 2]
```diff
@@ -48,6 +49,8 @@
 
   public static final String ACCEPT_VERSION_HEADER_NAME = "Accept-Version";
   public static final String SUPPORTED_METADATA_TYPE = "cst";
+  private static final long GIGABYTE = 1 << 30;
+  public static final long MAX_FILE_BYTES = 10 * GIGABYTE;
```
Comment (emmberk) [inline src/main/java/com/connexta/store/controllers/StoreController.java]: ✏️ `private`

File: src/main/java/com/connexta/store/controllers/StoreController.java [hunk 3]
```diff
@@ -69,19 +72,30 @@
     }
 
     final Long fileSize = file.getSize();
-    // TODO validate that fileSize is (0 GB, 10 GB]
+    if (fileSize > MAX_FILE_BYTES) {
+      throw new CreateDatasetException(
```
Comment (emmberk) [inline src/main/java/com/connexta/store/controllers/StoreController.java]: ✏️ `throw new ValidationException`

File: src/main/java/com/connexta/store/controllers/StoreController.java [hunk 4]
```diff
@@ -69,19 +72,30 @@
...
+          BAD_REQUEST,
+          String.format(
+              "File size is %d bytes. File size cannot be greater than %d bytes",
+              fileSize, MAX_FILE_BYTES));
+    }
 
     final String mediaType = file.getContentType();
-    // TODO verify that mediaType is not blank and is a valid Content Type
+    if (mediaType == null) {
```
Comment (emmberk) [inline src/main/java/com/connexta/store/controllers/StoreController.java]: ✏️  If the mediaType is not null could it still be invalid? If so, I think it should be validated here, or we should keep that TODO.

File: src/main/java/com/connexta/store/controllers/StoreController.java [hunk 5]
```diff
@@ -69,19 +72,30 @@
...
+      throw new CreateDatasetException(BAD_REQUEST, "Media type is missing");
```
Comment (emmberk) [inline src/main/java/com/connexta/store/controllers/StoreController.java]: ✏️ `throw new ValidationException`

File: src/main/java/com/connexta/store/controllers/StoreController.java [hunk 6]
```diff
@@ -69,19 +72,30 @@
...
+    }
 
     final String fileName = file.getOriginalFilename();
-    // TODO verify that fileName is not blank
+    if (StringUtils.isEmpty(fileName)) {
+      throw new CreateDatasetException(BAD_REQUEST, "Filename is missing");
```
Comment (emmberk) [inline src/main/java/com/connexta/store/controllers/StoreController.java]: ✏️ `throw new ValidationException`

File: src/main/java/com/connexta/store/controllers/StoreController.java [hunk 7]
```diff
@@ -69,19 +72,30 @@
...
 
     final String mediaType = file.getContentType();
-    // TODO verify that mediaType is not blank and is a valid Content Type
+    if (mediaType == null) {
+      throw new CreateDatasetException(BAD_REQUEST, "Media type is missing");
+    }
 
     final String fileName = file.getOriginalFilename();
-    // TODO verify that fileName is not blank
+    if (StringUtils.isEmpty(fileName)) {
```
Comment (emmberk) [inline src/main/java/com/connexta/store/controllers/StoreController.java]: ✏️ `StringUtils.isBlank`

File: src/main/java/com/connexta/store/controllers/StoreController.java [hunk 8]
```diff
@@ -69,19 +72,30 @@
...
+      throw new CreateDatasetException(BAD_REQUEST, "Filename is missing");
+    }
 
     final InputStream inputStream;
     try {
       inputStream = file.getInputStream();
     } catch (IOException e) {
       throw new CreateDatasetException(
```
Comment (emmberk) [inline src/main/java/com/connexta/store/controllers/StoreController.java]: ✏️ `throw new ValidationException`

File: src/test/java/com/connexta/store/CreateDatasetTests.java [hunk 9]
```diff
@@ -18,28 +25,34 @@
...
+import org.springframework.web.multipart.MultipartFile;
+import org.springframework.web.server.ResponseStatusException;
 
 @SpringBootTest
 @AutoConfigureMockMvc
 public class CreateDatasetTests {
 
   @MockBean private AmazonS3 mockAmazonS3;
 
+  @Inject private StoreController storeController;
```
Comment (emmberk) [inline src/test/java/com/connexta/store/CreateDatasetTests.java]: ✏️ Tests in this class are using MockMvc. I think that tests that test StoreController directly belong in a different class.

File: src/test/java/com/connexta/store/CreateDatasetTests.java [hunk 10]
```diff
@@ -89,9 +102,40 @@ public void testBadAcceptVersion() throws Exception {
...
-    // TODO verify 400
+  void testCantReadAttachment() throws IOException {
+    MultipartFile file = spy(newValidMultipartFile());
+    doThrow(new IOException("Cannot read attachment")).when(file).getInputStream();
+    assertBadRequest(file);
+    verify(file).getInputStream();
+  }
+
+  @Test
+  void testFileTooLarge() {
```
Comment (peterhuffer) [inline src/test/java/com/connexta/store/CreateDatasetTests.java]: ❓ What happens when a negative file size is passed in?

File: src/main/java/com/connexta/store/controllers/StoreController.java (hunk 6 above)
Comment (ahoffer) [reply src/main/java/com/connexta/store/controllers/StoreController.java]: ValidateException. That would have been good to know before I got started. Let's put hints, tips, design constraints, or technical expectations into the comments on tickets.

File: src/main/java/com/connexta/store/controllers/StoreController.java (hunk 7 above)
Comment (peterhuffer) [inline src/main/java/com/connexta/store/controllers/StoreController.java]: ❓ Do we need to do any sort of escaping on these inputs?

Comment (ahoffer) [reply src/main/java/com/connexta/store/controllers/StoreController.java]: Not that I know of.

File: src/test/java/com/connexta/store/CreateDatasetTests.java (hunk 10 above)
Comment (ahoffer) [reply src/test/java/com/connexta/store/CreateDatasetTests.java]: Size comes from the stream and streams cannot have a negative length.

File: src/main/java/com/connexta/store/controllers/StoreController.java (hunk 4 above)
Comment (ahoffer) [reply src/main/java/com/connexta/store/controllers/StoreController.java]: AFAIK we except any media type. At this stage in the application, the File is just a stream of bits. We store it and ask transform to interpret it. If transform can interpret the File enough to know if it can be promoted, then all is well.

Comment (ahoffer) [reply src/main/java/com/connexta/store/controllers/StoreController.java]: I'll go ahead and add a test for empty or blank strings, but I think they are unlikely. These kinds of things are either null or well-formed. Usually.
